hydroflow/src/compiled/pull.rs

```diff
@@ -97,9 +121,83 @@ where
     }
 }
 
+pub struct CrossJoin<I1, V1, I2, V2>
+where
+    V1: Eq + Clone,
+    V2: Eq + Clone,
+    I1: Iterator<Item = V1>,
+    I2: Iterator<Item = V2>,
+{
+    lhs: I1,
+    rhs: I2,
+    state: CrossJoinState<V1, V2>,
+}
+
+impl<I1, V1: 'static, I2, V2: 'static> Iterator for CrossJoin<I1, V1, I2, V2>
+where
+    V1: Eq + Clone,
+    V2: Eq + Clone,
+    I1: Iterator<Item = V1>,
+    I2: Iterator<Item = V2>,
+{
+    type Item = (V1, V2);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            match self.state.eventside {
+                Side::Left => {
+                    if let Some(i) = self.state.opposite_ix.next() {
+                        // TODO(justin): unnecessary clone (sometimes).
+                        let l = self.state.ltab.last().unwrap().clone();
+                        let r = self.state.rtab.get(i).unwrap().clone();
+                        return Some((l, r));
+                    } 
+                }
+                Side::Right => {
+                    if let Some(i) = self.state.opposite_ix.next() {
+                        // TODO(justin): unnecessary clone (sometimes).
+                        let l = self.state.ltab.get(i).unwrap().clone();
+                        let r = self.state.rtab.last().unwrap().clone();
+                        return Some((l, r));
+                    }
+                } 
+            }
+
+            // IF WE'RE HERE THERE'S NO CURRENT ITERATION.
```

Comment:
nit: we usually do normal capitalization/punctuation in comments

Reply:
got it!